File: sdk/ruby/lib/rocksky/resources/artist.rb
module Rocksky
  module Resources
    # `app.rocksky.artist.*` endpoints.
    class Artist < Base
      # Fetch an artist by AT-URI.
      def get_artist(uri:)
        query("app.rocksky.artist.getArtist", uri: uri)
      end

      # List artists. `names` may be an Array of strings; it will be joined CSV-style.
      def get_artists(limit: nil, offset: nil, names: nil, genre: nil)
        query("app.rocksky.artist.getArtists",
              limit: limit, offset: offset, names: names, genre: genre)
      end

      # Albums by an artist.
      def get_artist_albums(uri:)
        query("app.rocksky.artist.getArtistAlbums", uri: uri)
      end

      # Tracks by an artist.
      def get_artist_tracks(uri:, limit: nil, offset: nil)
        query("app.rocksky.artist.getArtistTracks",
              uri: uri, limit: limit, offset: offset)
      end

      # All-time listeners for an artist.
      def get_artist_listeners(uri:, limit: nil, offset: nil)
        query("app.rocksky.artist.getArtistListeners",
              uri: uri, limit: limit, offset: offset)
      end

      # Recent listeners for an artist.
      def get_artist_recent_listeners(uri:, limit: nil, offset: nil)
        query("app.rocksky.artist.getArtistRecentListeners",
              uri: uri, limit: limit, offset: offset)
      end
    end
  end
end

File: sdk/python/src/rocksky/resources/artist.py
"""``app.rocksky.artist.*`` — artist views and per-artist data."""

from __future__ import annotations

from ..models import (
    AlbumBasic,
    Artist,
    ArtistBasic,
    ArtistListener,
    RecentListener,
    SongBasic,
)
from ._base import Resource, parse_list, parse_model


class ArtistResource(Resource):
    async def get(self, uri: str) -> Artist:
        data = await self._transport.query(
            "app.rocksky.artist.getArtist", params={"uri": uri}
        )
        return parse_model(Artist, data)

    async def list(
        self,
        *,
        limit: int | None = None,
        offset: int | None = None,
        names: list[str] | str | None = None,
        genre: str | None = None,
    ) -> list[ArtistBasic]:
        data = await self._transport.query(
            "app.rocksky.artist.getArtists",
            params={
                "limit": limit,
                "offset": offset,
                "names": names,
                "genre": genre,
            },
        )
        return parse_list(ArtistBasic, data, key="artists")

    async def get_albums(self, uri: str) -> list[AlbumBasic]:
        data = await self._transport.query(
            "app.rocksky.artist.getArtistAlbums", params={"uri": uri}
        )
        return parse_list(AlbumBasic, data, key="albums")

    async def get_tracks(
        self,
        uri: str,
        *,
        limit: int | None = None,
        offset: int | None = None,
    ) -> list[SongBasic]:
        data = await self._transport.query(
            "app.rocksky.artist.getArtistTracks",
            params={"uri": uri, "limit": limit, "offset": offset},
        )
        return parse_list(SongBasic, data, key="tracks")

    async def get_listeners(
        self,
        uri: str,
        *,
        limit: int | None = None,
        offset: int | None = None,
    ) -> list[ArtistListener]:
        data = await self._transport.query(
            "app.rocksky.artist.getArtistListeners",
            params={"uri": uri, "limit": limit, "offset": offset},
        )
        return parse_list(ArtistListener, data, key="listeners")

    async def get_recent_listeners(
        self,
        uri: str,
        *,
        limit: int | None = None,
        offset: int | None = None,
    ) -> list[RecentListener]:
        data = await self._transport.query(
            "app.rocksky.artist.getArtistRecentListeners",
            params={"uri": uri, "limit": limit, "offset": offset},
        )
        return parse_list(RecentListener, data, key="listeners")
